src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx

```diff
@@ -1211,16 +1227,63 @@ namespace xclcpuemhal2 {
       return;
     }
 
+    setDeviceProcessStarted(false);
     std::string socketName = sock->get_name();
     if(socketName.empty() == false)// device is active if socketName is non-empty
     {
 #ifndef _WINDOWS
       xclClose_RPC_CALL(xclClose,this);
 #endif
     }
+   closemMessengerThread();
    saveDeviceProcessOutput();
   }
 
+  void CpuemShim::closemMessengerThread()
+  {
+    if (mMessengerThreadStarted) {
+      mMessengerThread.join();
+      mMessengerThreadStarted = false;
+    }
+  }
+
+  void messagesThread(xclcpuemhal2::CpuemShim *inst)
+  {
+    static auto start_time = std::chrono::high_resolution_clock::now();
+    unsigned int timeCheck = 0;
+    unsigned int parseCount = 0;
+    while (inst && inst->getDeviceProcessStarted())
+    {
+      auto end_time = std::chrono::high_resolution_clock::now();
+      if (std::chrono::duration<double>(end_time - start_time).count() > timeCheck) {
+        start_time = std::chrono::high_resolution_clock::now();
+        inst->parseLog();
+        parseCount++;
+        if (parseCount%5 == 0 && timeCheck < 300)
+          timeCheck += 10;
+      }
+    }
+  }
+
+  int CpuemShim::parseLog()
+  {
+    std::string matchString = "received request to end simulation from connected initiator";
+    std::ifstream ifs(deviceDirectory + "/../../../device_process.log");
+    if (ifs.is_open()) {
+      std::string line;
+      while (std::getline(ifs, line)) {
+        if (line.find(matchString) != std::string::npos) {
+          if(std::find(parsedMsgs.begin(), parsedMsgs.end(), line) == parsedMsgs.end()) {
+            std::cout << "Received request to end the application. Press Cntrl+C to exit the application." << std::endl;
+            parsedMsgs.push_back(line);
+            return 0;
+          }
+        }
+      }
```

Comment:
Too convoluted. Please check https://en.cppreference.com/w/cpp/string/basic_string/getline.

Reply:
This is removed. We have a parseLog structure which parses filecontent and prints as per matchString.